### PR TITLE
refactor(color-picker)!: simplify color data structure

### PR DESCRIPTION
**Pull request by Walkeezy**
This PR simplifies the color data structure as described in the following ticket, to avoid having colors in multiple color spaces (rgb, hex etc.): https://app.clickup.com/t/22b6p20

It also removes most of the utility methods in `src/utilities/colors.ts` and uses the `tinycolor2` library for color manipulation and conversion instead.

### CODE REVIEW

File: src/components/ColorPicker/BrandColorPicker.tsx [hunk 1]
```diff
@@ -81,18 +84,22 @@ export const BrandColorPicker: FC<Props> = ({ palettes: defaultPalettes = [], cu
...
+                                                      isColorLight(color) ? "tw-text-black" : "tw-text-white"
+                                                  }`}
+                                                  style={{ background: tinycolor(color).toRgbString() }}
                                               >
-                                                  {color.hex === currentColor.hex && (
-                                                      <IconCheck size={IconSize.Size20} />
-                                                  )}
+                                                  {color.r === currentColor.r &&
+                                                      color.g === currentColor.g &&
+                                                      color.b === currentColor.b && (
```
Comment (hochreutenerl) [inline src/components/ColorPicker/BrandColorPicker.tsx:100]: Probably we should consider `a` too here

File: src/components/ColorPicker/CustomColorPicker.tsx [hunk 2]
```diff
@@ -32,21 +30,25 @@ export const CustomColorPicker: FC<Omit<ColorPickerProps, "palette">> = ({
...
-        setHexInput(hex.substring(1));
-    }, [hex]);
+        setHexInput(parsedColor.toHex());
+    }, [color]);
 
     const handleHexChange = () => {
-        if (isValidHex(hexInput)) {
-            onSelect(toColor(currentColor, { hex: hexInput }));
+        const parsedHex = tinycolor(hexInput);
+        parsedHex.isValid();
```
Comment (hochreutenerl) [inline src/components/ColorPicker/CustomColorPicker.tsx:49]: This line is duplicated and can be removed

File: src/utilities/colors.ts [hunk 3]
```diff
@@ -1,100 +1,62 @@
 /* (c) Copyright Frontify Ltd., all rights reserved. */
 
-/* eslint-disable @typescript-eslint/ban-ts-comment */
```
Comment (hochreutenerl) [inline src/utilities/colors.ts:3]: Really nice that we can clean up a lot here.

File: src/components/ColorPicker/ColorPicker.tsx [hunk 4]
```diff
@@ -70,21 +69,19 @@ export const ColorPicker: FC<ColorPickerProps> = ({
...
-            style={{ background: backgroundColor, color: labelColor }}
+            className={`tw-flex tw-flex-none tw-justify-center tw-p-7 tw-text-m tw-gap-2 ${
+                isColorLight(color) ? "tw-text-black" : "tw-text-white"
+            }`}
+            style={{ background: backgroundColor }}
             data-test-id="color-preview"
         >
-            {name && <span className="tw-font-bold">{name}</span>}
-            <span className={name ? "" : "tw-font-bold"}>{displayValue}</span>
+            {color.name && <span className="tw-font-bold">{color.name}</span>}
```
Comment (NadiaPosch) [inline src/components/ColorPicker/ColorPicker.tsx:83]: You could create a `const colorName = color.name` above and reuse it

File: src/components/ColorPicker/CustomColorPicker.tsx [hunk 5]
```diff
@@ -32,21 +30,25 @@ export const CustomColorPicker: FC<Omit<ColorPickerProps, "palette">> = ({
...
+        setColor(currentColor);
     }, [currentColor]);
 
-    const [hexInput, setHexInput] = useState(hex.substring(1));
+    const [hexInput, setHexInput] = useState(parsedColor.toHex());
 
     useEffect(() => {
-        setHexInput(hex.substring(1));
-    }, [hex]);
+        setHexInput(parsedColor.toHex());
```
Comment (NadiaPosch) [inline src/components/ColorPicker/CustomColorPicker.tsx:44]: You could move this inside the `useEffect` from above since if the `currentColor` changes, so does `color`

File: src/components/ColorPicker/CustomColorPicker.tsx (hunk 2 above)
Comment (NadiaPosch) [inline src/components/ColorPicker/CustomColorPicker.tsx:49]: Please remove if not needed

File: src/components/ColorPicker/CustomColorPicker.tsx [hunk 6]
```diff
@@ -55,10 +57,11 @@ export const CustomColorPicker: FC<Omit<ColorPickerProps, "palette">> = ({
             <div className="tw-flex tw-gap-2 tw-w-full tw-h-[200px]">
                 <div className="tw-relative tw-flex-grow tw-overflow-hidden tw-rounded">
                     <Saturation
-                        hsl={hsl}
-                        hsv={hsv}
+                        rgb={parsedColor.toRgb()}
+                        hsl={parsedColor.toHsl()}
+                        hsv={parsedColor.toHsv()}
```
Comment (NadiaPosch) [inline src/components/ColorPicker/CustomColorPicker.tsx:62]: You could create variables `rgb, hsl, hsv` and reuse below

File: src/components/ColorPicker/CustomColorPicker.tsx [hunk 7]
```diff
@@ -55,10 +57,11 @@ export const CustomColorPicker: FC<Omit<ColorPickerProps, "palette">> = ({
...
                         pointer={() => <ColorPointer />}
-                        onChange={debounce((color) => onSelect(toColor(currentColor, { hex: toState(color).hex })))}
+                        onChange={debounce((color) => onSelect(tinycolor(color).toRgb()))}
```
Comment (NadiaPosch) [inline src/components/ColorPicker/CustomColorPicker.tsx:64]: You could create an `onChangeColor` method and reuse below

File: src/components/ColorInputFlyout/ColorPickerFlyout.spec.tsx [hunk 8]
```diff
@@ -45,10 +46,10 @@ describe("ColorInputFlyout Component", () => {
...
     it("should render with initial color", () => {
-        mount(<Component currentColor={{ hex: TEST_COLOR.hex, alpha: 1 }} />);
+        mount(<Component currentColor={TEST_COLOR} />);
 
-        cy.get(TRIGGER_ID).should("contain", TEST_COLOR.hex).click();
-        cy.get(COLOR_PREVIEW_ID).should("contain", TEST_COLOR.hex);
-        cy.get(MENU_ITEM_DECORATOR_ID).children("span").invoke("attr", "style").should("include", TEST_COLOR.rgb);
+        cy.get(TRIGGER_ID).should("contain", TEST_COLOR_HEX).click();
+        cy.get(COLOR_PREVIEW_ID).should("contain", TEST_COLOR_HEX);
+        cy.get(MENU_ITEM_DECORATOR_ID).children("span").invoke("attr", "style").should("include", "rgb(0, 133, 255)");
```
Comment (NadiaPosch) [inline src/components/ColorInputFlyout/ColorPickerFlyout.spec.tsx:53]:
```suggestion
        cy.get(MENU_ITEM_DECORATOR_ID).children("span").invoke("attr", "style").should("include", `rgb(${Object.values(TEST_COLOR).join(",")})`);
```

**Review by hochreutenerl — CHANGES_REQUESTED** [review]
Nice work, I think this already simplifies the colors.
Please check the annotations as there are still some things we need to have a second look first.

File: src/utilities/colors.ts [hunk 9]
```diff
@@ -1,100 +1,62 @@
...
+    const palette = [...new Array(amount)].map((_, index) => {
+        const name = (90 - index * 10).toString();
+        const { r, g, b, a } = sourceColor.lighten(index * 3).toRgb();
+        return {
+            r,
+            g,
+            b,
+            a,
+            name,
+        };
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/utilities/colors.ts:36]: Nice, changed

File: src/components/ColorPicker/CustomColorPicker.tsx (hunk 2 above)
Comment (Walkeezy) [reply src/components/ColorPicker/CustomColorPicker.tsx:49]: Removed

Comment (Walkeezy) [reply src/components/ColorPicker/CustomColorPicker.tsx:49]: Removed

File: src/components/ColorInputFlyout/ColorPickerFlyout.spec.tsx (hunk 8 above)
Comment (Walkeezy) [reply src/components/ColorInputFlyout/ColorPickerFlyout.spec.tsx:53]: Changed

File: src/components/ColorPicker/ColorPicker.tsx (hunk 4 above)
Comment (Walkeezy) [reply src/components/ColorPicker/ColorPicker.tsx:83]: Added

File: src/components/ColorPicker/CustomColorPicker.tsx (hunk 5 above)
Comment (Walkeezy) [reply src/components/ColorPicker/CustomColorPicker.tsx:44]: Changed

File: src/components/ColorPicker/CustomColorPicker.tsx (hunk 6 above)
Comment (Walkeezy) [reply src/components/ColorPicker/CustomColorPicker.tsx:62]: Added

File: src/components/ColorPicker/BrandColorPicker.tsx (hunk 1 above)
Comment (Walkeezy) [reply src/components/ColorPicker/BrandColorPicker.tsx:100]: Good catch, added 👍🏻

File: src/components/ColorPicker/CustomColorPicker.tsx (hunk 7 above)
Comment (Walkeezy) [reply src/components/ColorPicker/CustomColorPicker.tsx:62]: Would rather keep it like this, it's short enough and easier to understand what's going on IMO.

File: src/utilities/colors.ts [hunk 10]
```diff
@@ -1,100 +1,62 @@
...
+    });
+    return palette;
+};
+
 export const EXAMPLE_PALETTES: Palette[] = [
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/utilities/colors.ts:38]: Good input, I agree that the `EXAMPLE_PALETTES` not really belong to the colors utility. I've moved it to `ColorPicker.stories.tsx` now.
I'm aware that the values are different now, but that shouldn't really matter. :)

File: src/components/ColorInputFlyout/ColorPickerTrigger.tsx [hunk 11]
```diff
@@ -49,7 +50,7 @@ export const ColorInputTrigger: FC<ColorInputTriggerProps> = ({
                                     "tw-h-4 tw-w-4 tw-rounded tw-flex tw-items-center tw-justify-center",
                                     disabled && "tw-opacity-50",
                                 ])}
-                                style={{ background: getBackgroundColor(currentColor) }}
+                                style={{ background: backgroundColor }}
```
Comment (SamuelAlev) [inline src/components/ColorInputFlyout/ColorPickerTrigger.tsx:53]:
```suggestion
                                style={{ backgroundColor }}
```

File: src/components/ColorPicker/ColorPicker.stories.tsx [hunk 12]
```diff
@@ -1,17 +1,53 @@
...
+        const lightColor = sourceColor.lighten(index * 3).toRgb();
+        return {
+            ...lightColor,
+            name,
+        };
+    });
+    return palette;
+};
+
+export const EXAMPLE_PALETTES: Palette[] = [
```
Comment (SamuelAlev) [inline src/components/ColorPicker/ColorPicker.stories.tsx:22]:
```suggestion
const EXAMPLE_PALETTES: Palette[] = [
```
Otherwise on Storybook 
Capture d’écran 2022-01-21 à 09 59 40

File: src/foundation/Icon/Generated/IconTwitter.tsx [hunk 13]
```diff
@@ -0,0 +1,27 @@
+import React, { memo } from "react";
```
Comment (SamuelAlev) [inline src/foundation/Icon/Generated/IconTwitter.tsx:1]: Coul you rebase on `main`?

**Review by SamuelAlev — COMMENTED** [review]
nice one! :D

File: src/components/ColorPicker/ColorPicker.tsx [hunk 14]
```diff
@@ -70,21 +69,20 @@ export const ColorPicker: FC<ColorPickerProps> = ({
 };
 
 const ColorPreview: FC<{ color: Color; format: ColorFormat }> = ({ color, format }) => {
-    const { hex, rgba, name, alpha } = color;
-    const backgroundColor = getBackgroundColor(color);
+    const parsedColor = tinycolor(color);
+    const backgroundColor = parsedColor.toRgbString();
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/components/ColorPicker/ColorPicker.tsx:73]: The name is a little bit misleading, it actually has the alpha channel: https://github.com/bgrins/TinyColor#torgbstring :)

File: src/components/ColorPicker/ColorPicker.stories.tsx (hunk 12 above)
Comment (Walkeezy) [reply src/components/ColorPicker/ColorPicker.stories.tsx]: Moved it to its own file now :)

File: src/foundation/Icon/Generated/IconTwitter.tsx (hunk 13 above)
Comment (Walkeezy) [reply src/foundation/Icon/Generated/IconTwitter.tsx:1]: Done

File: src/components/ColorPicker/ColorPicker.tsx [hunk 15]
```diff
@@ -70,21 +69,20 @@ export const ColorPicker: FC<ColorPickerProps> = ({
...
-    }, [hex, rgba, alpha]);
-
     return (
         <div
-            className="tw-flex tw-flex-none tw-justify-center tw-p-7 tw-text-m tw-text-black dark:tw-text-white tw-gap-2"
-            style={{ background: backgroundColor, color: labelColor }}
+            className={`tw-flex tw-flex-none tw-justify-center tw-p-7 tw-text-m tw-gap-2 ${
+                isColorLight(color) ? "tw-text-black" : "tw-text-white"
+            }`}
+            style={{ background: backgroundColor }}
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/components/ColorPicker/ColorPicker.tsx]: Niice, changed

File: src/components/ColorInputFlyout/ColorPickerTrigger.tsx (hunk 11 above)
Comment (Walkeezy) [reply src/components/ColorInputFlyout/ColorPickerTrigger.tsx:53]: Changed

File: src/components/ColorPicker/CustomColorPicker.tsx [hunk 16]
```diff
@@ -103,8 +106,8 @@ export const CustomColorPicker: FC<Omit<ColorPickerProps, "palette">> = ({
                             value={hexInput}
                             decorator="#"
                             size={6}
-                            onChange={(hex) => {
-                                setHexInput(hex);
+                            onChange={(input) => {
+                                setHexInput(input);
                             }}
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/components/ColorPicker/CustomColorPicker.tsx:111]: Changed

File: src/components/ColorPicker/ColorPicker.tsx [hunk 17]
```diff
@@ -69,22 +68,24 @@ export const ColorPicker: FC<ColorPickerProps> = ({
...
-        <div
-            className="tw-flex tw-flex-none tw-justify-center tw-p-7 tw-text-m tw-text-black dark:tw-text-white tw-gap-2"
-            style={{ background: backgroundColor, color: labelColor }}
-            data-test-id="color-preview"
-        >
-            {name && <span className="tw-font-bold">{name}</span>}
-            <span className={name ? "" : "tw-font-bold"}>{displayValue}</span>
+        <div className="tw-sticky tw-top-0 tw-bg-white tw-z-20 dark:tw-bg-black-95">
+            <div
+                className={`tw-flex tw-justify-center tw-p-7 tw-text-m tw-gap-2 ${
```
Comment (NadiaPosch) [inline src/components/ColorPicker/ColorPicker.tsx:79]: You could use `merge` here as well

File: src/components/ColorPicker/CustomColorPicker.tsx (hunk 7 above)
Comment (NadiaPosch) [reply src/components/ColorPicker/CustomColorPicker.tsx:62]: okay 👍

File: src/components/ColorPicker/BrandColorPicker.tsx [hunk 18]
```diff
@@ -81,18 +84,23 @@ export const BrandColorPicker: FC<Props> = ({ palettes: defaultPalettes = [], cu
...
                                               className="tw-flex tw-overflow-hidden tw-w-full"
-                                              onClick={() => onSelect(toColor(color, color))}
+                                              onClick={() => onSelect(color)}
                                           >
                                               <span
-                                                  className="tw-h-8 tw-w-8 tw-mr-2 tw-rounded tw-flex tw-items-center tw-justify-center tw-text-white"
-                                                  style={{ background: color.hex }}
+                                                  className={`tw-h-8 tw-w-8 tw-mr-2 tw-rounded tw-flex tw-items-center tw-justify-center ${
+                                                      isColorLight(color) ? "tw-text-black" : "tw-text-white"
+                                                  }`}
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/components/ColorPicker/BrandColorPicker.tsx:95]: Changed

File: src/components/ColorInputFlyout/ColorPickerFlyout.tsx [hunk 19]
```diff
@@ -30,9 +31,9 @@ export const ColorPickerFlyout: FC<ColorPickerFlyoutProps> = ({
         <Flyout
             hug={false}
             onClick={onClick}
-            scrollable={false}
             onClose={onClose}
             isOpen={open}
+            fixedHeader={<ColorPreview color={currentColor || { r: 255, g: 255, b: 255 }} format={currentFormat} />}
```
Comment (NadiaPosch) [inline src/components/ColorInputFlyout/ColorPickerFlyout.tsx:36]: This was already fixed in: #549 
Please rebase

Comment (Walkeezy) [reply src/components/ColorInputFlyout/ColorPickerFlyout.tsx:36]: That's from the rebase, but it had conflicts :)

File: src/components/ColorPicker/ColorPicker.tsx (hunk 17 above)
Comment (Walkeezy) [reply src/components/ColorPicker/ColorPicker.tsx]: Done

File: src/components/ColorPicker/BrandColorPicker.tsx [hunk 20]
```diff
@@ -81,18 +84,23 @@ export const BrandColorPicker: FC<Props> = ({ palettes: defaultPalettes = [], cu
...
+                                                  style={{ background: tinycolor(color).toRgbString() }}
                                               >
-                                                  {color.hex === currentColor.hex && (
-                                                      <IconCheck size={IconSize.Size20} />
-                                                  )}
+                                                  {color.r === currentColor.r &&
+                                                      color.g === currentColor.g &&
+                                                      color.b === currentColor.b &&
+                                                      color.a === currentColor.a && (
```
[the remark this answers is not in the document]
Comment (Walkeezy) [reply src/components/ColorPicker/BrandColorPicker.tsx:102]: We are not using `lodash` on arcade yet 🤔

**Review by NadiaPosch — APPROVED** [review]

File: src/components/ColorPicker/BrandColorPicker.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (SamuelAlev) [reply src/components/ColorPicker/BrandColorPicker.tsx:102]: Whaaat, nice well then it is good 👍

**Review by SamuelAlev — APPROVED** [review]

File: package.json [hunk 21]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@frontify/arcade",
-    "version": "0.16.2",
+    "version": "0.17.0",
```
Comment (hochreutenerl) [inline package.json:3]: Note that we have breaking changes on the interface. Currently I am not yet familiar on how to handle them between the repositories. (The default for semantic versioning would be to increase the major version)

File: src/foundation/Icon/Svg/Twitter.svg [hunk 22]
```diff
@@ -0,0 +1,3 @@
+<svg viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (hochreutenerl) [inline src/foundation/Icon/Svg/Twitter.svg:1]: Looks like there are changes in this PR which are not related to the topic

File: package.json (hunk 21 above)
Comment (SamuelAlev) [reply package.json:3]: Should be taken care of outside of PR when creating the release

File: src/foundation/Icon/Svg/Twitter.svg (hunk 22 above)
Comment (Walkeezy) [reply src/foundation/Icon/Svg/Twitter.svg:1]: I needed to rebase and had conflicts. Guess those commits are added to the history then.

Comment (hochreutenerl) [reply src/foundation/Icon/Svg/Twitter.svg:1]: The commits might be still there, but the changes against master here should disappear again.

Comment (Walkeezy) [reply src/foundation/Icon/Svg/Twitter.svg:1]: Just noticed I made a bit of a mess, fixing it right now, sorry 😬

**Review by hochreutenerl — APPROVED** [review]
All fine. 👍

File: src/foundation/Icon/Svg/Twitter.svg (hunk 22 above)
Comment (hochreutenerl) [reply src/foundation/Icon/Svg/Twitter.svg]: No problem. Now it looks good.
